commands/builds/create.js

```diff
@@ -8,84 +8,93 @@ let path = require('path')
 let exec = require('child_process').execSync
 let nodeTar = require('../../lib/node_tar')
 
-function compressSource (context, cwd, tempFile, cb) {
-  var tar = context.flags['tar'] || 'tar'
-  let tarVersion = exec(tar + ' --version').toString()
+function compressSource (context, cwd, tempFile) {
+  return new Promise(function (resolve, reject) {
+    var tar = context.flags['tar'] || 'tar'
+    let tarVersion = ''
 
-  if (tarVersion.match(/GNU tar/)) {
-    let includeVcsIgnore = context.flags['include-vcs-ignore']
-    let command = tar + ' cz -C ' + cwd + ' --exclude .git --exclude .gitmodules .'
-
-    if (!includeVcsIgnore) {
-      command += ' --exclude-vcs-ignores'
+    try {
+      tarVersion = exec(tar + ' --version').toString()
+    } catch (err) {
+      // tar was not found on the system
     }
 
-    exec(command + ' > ' + tempFile)
-    cb()
-  } else {
-    cli.warn('Couldn\'t detect GNU tar. Builds could fail due to decompression errors')
-    cli.warn('See https://devcenter.heroku.com/articles/platform-api-deploying-slugs#create-slug-archive')
-    cli.warn('Please install it, or specify the \'--tar\' option')
-    cli.warn('Falling back to node\'s built-in compressor')
-    nodeTar.call(cwd, tempFile, cb)
-  }
-}
+    if (tarVersion.match(/GNU tar/)) {
+      let includeVcsIgnore = context.flags['include-vcs-ignore']
+      let command = tar + ' cz -C ' + cwd + ' --exclude .git --exclude .gitmodules .'
 
-function uploadCwdToSource (context, heroku, cwd, fn) {
-  let tempFilePath = path.join(os.tmpdir(), uuid.v4() + '.tar.gz')
+      if (!includeVcsIgnore) {
+        command += ' --exclude-vcs-ignores'
+      }
 
-  heroku.request({
-    method: 'POST',
-    path: '/sources'
-  }).then(function (source) {
-    compressSource(context, cwd, tempFilePath, function () {
-      var stream = fs.createReadStream(tempFilePath)
-      stream.on('close', function () {
-        fs.unlink(tempFilePath)
-      })
+      exec(command + ' > ' + tempFile)
+      resolve()
+    } else {
+      cli.warn('Couldn\'t detect GNU tar. Builds could fail due to decompression errors')
+      cli.warn('See https://devcenter.heroku.com/articles/platform-api-deploying-slugs#create-slug-archive')
+      cli.warn('Please install it, or specify the \'--tar\' option')
+      cli.warn('Falling back to node\'s built-in compressor')
+      nodeTar.call(cwd, tempFile).then(resolve, reject)
+    }
+  })
+}
 
-      cli.got.put(source.source_blob.put_url, {
-        body: stream,
-        headers: {
-          'Content-Type': '',
-          'Content-Length': fs.statSync(tempFilePath).size
-        }
-      })
-        .then(function () {
-          fn(source.source_blob.get_url)
+function uploadCwdToSource (context, heroku, cwd) {
```

Comment:
this could be written better:

```js
async function uploadCwdToSource (context, heroku, cwd) {
  let tempFilePath = path.join(os.tmpdir(), uuid.v4() + '.tar.gz')
  let source = await heroku.post('/sources')
  await compressSource(context, cwd, tempFilePath)
  await cli.got.put(source.source_blob.put_url, {
    body: fs.createReadStream(tempFilePath),
    headers: {/*...*/}
  })
  return source.source_blob.get_url
}
```

note I haven't tested this at all